File: src/main/java/team4/footwithme/chat/service/ChatroomServiceImpl.java
package team4.footwithme.chat.service;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import team4.footwithme.chat.domain.Chatroom;
import team4.footwithme.chat.domain.ReservationChatroom;
import team4.footwithme.chat.domain.TeamChatroom;
import team4.footwithme.chat.repository.ChatroomRepository;
import team4.footwithme.chat.repository.RedisChatroomRepository;
import team4.footwithme.chat.service.request.ChatroomServiceRequest;
import team4.footwithme.chat.service.response.ChatroomResponse;
import team4.footwithme.global.exception.ExceptionMessage;

@Service
public class ChatroomServiceImpl implements ChatroomService {
    private final ChatroomRepository chatroomRepository;

    private final RedisChatroomRepository redisChatroomRepository;

    public ChatroomServiceImpl(ChatroomRepository chatroomRepository, RedisChatroomRepository redisChatroomRepository) {
        this.chatroomRepository = chatroomRepository;
        this.redisChatroomRepository = redisChatroomRepository;
    }

    /**
     * 채팅방 생성
     * 채팅방을 만든 사람은 자동으로 채팅방 초대
     *
     * @param request 채팅방 이름 ( 팀 or 예약 만들어질 때 이름 넣어서 보내주기 )
     * @return
     */
    @Transactional
    @Override
    public ChatroomResponse createChatroom(ChatroomServiceRequest request) {

        Chatroom chatroom = chatroomRepository.save(Chatroom.create(request.name()));

        // redis Hash에 저장
        redisChatroomRepository.createChatRoom(chatroom);

        return new ChatroomResponse(chatroom);
    }

    @Transactional
    @Override
    public ChatroomResponse createTeamChatroom(ChatroomServiceRequest request, Long teamId) {

        Chatroom chatroom = chatroomRepository.save(TeamChatroom.create(request.name(), teamId));

        // redis Hash에 저장
        redisChatroomRepository.createChatRoom(chatroom);

        return new ChatroomResponse(chatroom);
    }

    @Transactional
    @Override
    public ChatroomResponse createReservationChatroom(ChatroomServiceRequest request, Long reservationId) {

        Chatroom chatroom = chatroomRepository.save(ReservationChatroom.create(request.name(), reservationId));

        // redis Hash에 저장
        redisChatroomRepository.createChatRoom(chatroom);

        return new ChatroomResponse(chatroom);
    }

    /**
     * 채팅방 삭제 ( 채팅방 인원 삭제도 같이 진행해야 함 )
     *
     * @param chatroomId
     * @return
     */
    @Transactional
    @Override
    public Long deleteChatroomByChatroomId(Long chatroomId) {
        Chatroom chatroom = getChatroomByChatroomId(chatroomId);

        return deleteChatroom(chatroom);
    }

    @Transactional
    @Override
    public Long deleteTeamChatroom(Long teamId) {
        Chatroom chatroom = getChatroomByTeamId(teamId);

        return deleteChatroom(chatroom);
    }

    @Transactional
    @Override
    public Long deleteReservationChatroom(Long reservationId) {
        Chatroom chatroom = getChatroomByReservationId(reservationId);

        return deleteChatroom(chatroom);
    }

    /**
     * 채팅방 수정
     */
    @Transactional
    @Override
    public ChatroomResponse updateChatroom(Long chatroomId, ChatroomServiceRequest request) {
        Chatroom chatroom = getChatroomByChatroomId(chatroomId);

        chatroom.updateName(request.name());

        return new ChatroomResponse(chatroom);
    }

    private Long deleteChatroom(Chatroom chatroom) {
        redisChatroomRepository.deleteChatroomFromRedis(chatroom.getChatroomId());

        chatroomRepository.deleteById(chatroom.getChatroomId());
        return chatroom.getChatroomId();
    }

    private Chatroom getChatroomByChatroomId(Long chatroomId) {
        return chatroomRepository.findByChatroomId(chatroomId)
            .orElseThrow(() -> new IllegalArgumentException(ExceptionMessage.CHATROOM_NOT_FOUND.getText()));
    }

    private Chatroom getChatroomByTeamId(Long teamId) {
        return chatroomRepository.findByTeamId(teamId)
            .orElseThrow(() -> new IllegalArgumentException(ExceptionMessage.CHATROOM_NOT_FOUND.getText()));
    }

    private Chatroom getChatroomByReservationId(Long reservationId) {
        return chatroomRepository.findByReservationId(reservationId)
            .orElseThrow(() -> new IllegalArgumentException(ExceptionMessage.CHATROOM_NOT_FOUND.getText()));
    }
}

File: src/main/java/team4/footwithme/chat/service/ChatroomServiceImpl.kt
package team4.footwithme.chat.service

import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import team4.footwithme.chat.domain.Chatroom
import team4.footwithme.chat.domain.ReservationChatroom
import team4.footwithme.chat.domain.TeamChatroom
import team4.footwithme.chat.repository.ChatroomRepository
import team4.footwithme.chat.repository.RedisChatroomRepository
import team4.footwithme.chat.service.request.ChatroomServiceRequest
import team4.footwithme.chat.service.response.ChatroomResponse
import team4.footwithme.global.exception.ExceptionMessage

@Service
class ChatroomServiceImpl(
    private val chatroomRepository: ChatroomRepository,
    private val redisChatroomRepository: RedisChatroomRepository
) : ChatroomService {
    /**
     * 채팅방 생성
     * 채팅방을 만든 사람은 자동으로 채팅방 초대
     *
     * @param request 채팅방 이름 ( 팀 or 예약 만들어질 때 이름 넣어서 보내주기 )
     * @return
     */
    @Transactional
    override fun createChatroom(request: ChatroomServiceRequest?): ChatroomResponse {
        val chatroom = chatroomRepository.save<Chatroom>(Chatroom.Companion.create(request!!.name))

        // redis Hash에 저장
        redisChatroomRepository.createChatRoom(chatroom)

        return ChatroomResponse(chatroom)
    }

    @Transactional
    override fun createTeamChatroom(request: ChatroomServiceRequest, teamId: Long?): ChatroomResponse {
        val chatroom: Chatroom =
            chatroomRepository.save<TeamChatroom>(TeamChatroom.Companion.create(request.name, teamId))

        // redis Hash에 저장
        redisChatroomRepository.createChatRoom(chatroom)

        return ChatroomResponse(chatroom)
    }

    @Transactional
    override fun createReservationChatroom(request: ChatroomServiceRequest, reservationId: Long?): ChatroomResponse {
        val chatroom: Chatroom = chatroomRepository.save<ReservationChatroom>(
            ReservationChatroom.Companion.create(
                request.name,
                reservationId
            )
        )

        // redis Hash에 저장
        redisChatroomRepository.createChatRoom(chatroom)

        return ChatroomResponse(chatroom)
    }

    /**
     * 채팅방 삭제 ( 채팅방 인원 삭제도 같이 진행해야 함 )
     *
     * @param chatroomId
     * @return
     */
    @Transactional
    override fun deleteChatroomByChatroomId(chatroomId: Long?): Long? {
        val chatroom = getChatroomByChatroomId(chatroomId)

        return deleteChatroom(chatroom)
    }

    @Transactional
    override fun deleteTeamChatroom(teamId: Long): Long? {
        val chatroom = getChatroomByTeamId(teamId)

        return deleteChatroom(chatroom)
    }

    @Transactional
    override fun deleteReservationChatroom(reservationId: Long): Long? {
        val chatroom = getChatroomByReservationId(reservationId)

        return deleteChatroom(chatroom)
    }

    /**
     * 채팅방 수정
     */
    @Transactional
    override fun updateChatroom(chatroomId: Long?, request: ChatroomServiceRequest?): ChatroomResponse {
        val chatroom = getChatroomByChatroomId(chatroomId)

        chatroom!!.updateName(request!!.name)

        return ChatroomResponse(chatroom)
    }

    private fun deleteChatroom(chatroom: Chatroom?): Long? {
        redisChatroomRepository.deleteChatroomFromRedis(chatroom!!.chatroomId)

        chatroomRepository.deleteById(chatroom.chatroomId)
        return chatroom.chatroomId
    }

    private fun getChatroomByChatroomId(chatroomId: Long?): Chatroom? {
        return chatroomRepository.findByChatroomId(chatroomId)
            .orElseThrow { IllegalArgumentException(ExceptionMessage.CHATROOM_NOT_FOUND.text) }
    }

    private fun getChatroomByTeamId(teamId: Long): Chatroom? {
        return chatroomRepository.findByTeamId(teamId)
            .orElseThrow { IllegalArgumentException(ExceptionMessage.CHATROOM_NOT_FOUND.text) }
    }

    private fun getChatroomByReservationId(reservationId: Long): Chatroom? {
        return chatroomRepository.findByReservationId(reservationId)
            .orElseThrow { IllegalArgumentException(ExceptionMessage.CHATROOM_NOT_FOUND.text) }
    }
}
